synapse/lib/crypto/ecc.py

```diff
@@ -0,0 +1,221 @@
+            encoding=c_ser.Encoding.DER,
+            format=c_ser.PrivateFormat.PKCS8,
+            encryption_algorithm=c_ser.NoEncryption())
+
+    @staticmethod
+    def load(byts):
+        '''
+        Create a PriKey instance from DER/PKCS8 encoded bytes.
+
+        Args:
+            byts (bytes): Bytes to load
+
+        Returns:
+            PriKey: A new PubKey instance.
+        '''
+        return PriKey(c_ser.load_der_private_key(
+                byts,
+                password=None,
+                backend=default_backend()))
+
+class PubKey:
+    '''
+    A helper class for using ECC public keys.
+    '''
+
+    def __init__(self, publ):
+        self.publ = publ  # type: c_ec.EllipticCurvePublicKey
+
+    def dump(self):
+        '''
+        Get the public key bytes in DER/SubjectPublicKeyInfo format.
+
+        Returns:
+            bytes: The DER/SubjectPublicKeyInfo encoded public key.
+        '''
+        return self.publ.public_bytes(
+            encoding=c_ser.Encoding.DER,
+            format=c_ser.PublicFormat.SubjectPublicKeyInfo)
+
+    def verify(self, byts, sign):
+        '''
+        Verify the signature for the given bytes using the ECC
+        public key.
+
+        Args:
+            byts (bytes): The data bytes.
+            sign (bytes): The signature bytes.
+
+        Returns:
+            bool: True if the data was verified, False otherwise.
+        '''
+        try:
+            chosen_hash = c_hashes.SHA256()
+            hasher = c_hashes.Hash(chosen_hash, default_backend())
+            hasher.update(byts)
+            digest = hasher.finalize()
+            self.publ.verify(sign,
+                             digest,
+                             c_ec.ECDSA(c_utils.Prehashed(chosen_hash))
+                             )
+            return True
+        except InvalidSignature as e:
+            logger.exception('Error in publ.verify')
+            return False
+
+    def iden(self):
+        '''
+        Return a SHA256 hash for the public key (to be used as a GUID).
+
+        Returns:
+            str: The SHA256 hash of the public key bytes.
+        '''
+        return hashlib.sha256(self.dump()).hexdigest()
+
+    @staticmethod
+    def load(byts):
+        '''
+        Create a PubKey instance from DER/PKCS8 encoded bytes.
+
+        Args:
+            byts (bytes): Bytes to load
+
+        Returns:
+            PubKey: A new PubKey instance.
+        '''
+        return PubKey(c_ser.load_der_public_key(
+                byts,
+                backend=default_backend()))
+
+def doECDHE(sprku, spbkv, eprku, epbkv,
+            length=64,
+            salt=None,
+            info=None):
+    '''
+    Perform one side of an Ecliptic Curve Diffie Hellman Ephemeral key exchange.
+
+    Args:
+        sprku (PriKey): Static Private Key for U
+        spbkv (PubKey: Static Public Key for V
+        eprku (PriKey): Ephemeral Private Key for U
+        epbkv (PubKey): Ephemeral Public Key for V
```

Comment:
since there are so many of these and they're so similar, i could totally see making them something like:
```
sesspubkey
statprivkey
```

Reply:
I had lifted the names from (largely) the nist docs there. its a trivial change to make

Reply:
This wasn't mine, but this sticks closely with the terminology that NIST 800-56A uses. At the same time, I prefer longer names for parameters.

statprv_u, statpub_v, ephmprv_u, ephmpub_v?

Reply:
i'll update that